### ISSUE TITLE
Debugging django without '--noreload' does not work

### ISSUE THREAD

**Issue by DonJayamanne** [Microsoft/ptvsd#1241]
@wyattbiker commented on [Wed Feb 27 2019](https://github.com/Microsoft/vscode-python/issues/4589)

Commenting out the --noreload key in launch.json disables breakpoints.
As a side effect, when you click stop button, python remains loaded and must be pkilled
sudo lsof -i -P -n | grep LISTEN

```
python extension 2019.2.5433 (27 Feb 2019)

Version: 1.31.1
Commit: 1b8e8302e405050205e69b59abb3559592bb9e60
Date: 2019-02-12T02:19:29.629Z
Electron: 3.1.2
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Linux x64 4.15.0-43-generic
```

```json
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Django",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/manage.py",
            "args": [
                "runserver",
                // "--noreload",
                "--nothreading"
            ],
            "django": true
        },
```

---

@DonJayamanne commented on [Tue Mar 05 2019](https://github.com/Microsoft/vscode-python/issues/4589#issuecomment-469770080)

We are however in the process of working on adding support for live reloading, and you can try it out today in the current release.

Please make the following changes to the `launch.json`:
* Remove `--noreload` (as you have done aready)
* Add `"subProcess": true`
I.e. modify as follows:
```js
        {
            "name": "Django",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/manage.py",
            "args": [
                "runserver",
                "--nothreading"
            ],
            "django": true,
			"subProcess": true
        }
```

Once you start debugging, it will take a second or two to attach to the sub procceses, and you should be good to go.
Basically we're working on adding support for debugging multiple processes.

**Please let me know how this goes.**

---

@wyattbiker commented on [Tue Mar 12 2019](https://github.com/Microsoft/vscode-python/issues/4589#issuecomment-472064113)

```
Name: Python
Id: ms-python.python
Description: Linting, Debugging (multi-threaded, remote), Intellisense, code formatting, refactoring, unit tests, snippets, and more.
Version: 2019.2.5558
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-python.python


    "version": "0.2.0",
    "configurations": [
        {
            "name": "Django Reload",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/manage.py",
            "args": [
                "runserver",
                "--nothreading"
            ],
            "django": true,
			"subProcess": true
        },  
```

Hello, i copied what you provided above. It runs fine, but crashes when I try to make a change/save to views.py. Line 8 (breakpoint line) crashes immediately. Line 9 will crash if I continue.

The right panel is manage.py

![image](https://user-images.githubusercontent.com/930478/54215344-2531b380-44be-11e9-8a7c-dcb5ade366b0.png)

Here is the complete error dump

```
Exception has occurred: SystemExit
3
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/utils/autoreload.py", line 276, in reloader_thread
    sys.exit(3)  # force reload
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/utils/autoreload.py", line 304, in python_reloader
    pass
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/utils/autoreload.py", line 323, in main
    python_reloader(wrapped_main_func, args, kwargs)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/commands/runserver.py", line 102, in run
    autoreload.main(self.inner_run, None, options)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/commands/runserver.py", line 95, in handle
    self.run(**options)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/base.py", line 353, in execute
    output = self.handle(*args, **options)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/commands/runserver.py", line 60, in execute
    super().execute(*args, **options)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/base.py", line 333, in run_from_argv
    pass
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/standing/public_html/dj/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/home/standing/public_html/mysite/manage.py", line 15, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.7/runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "/usr/local/lib/python3.7/runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
```

**Comment by karthiknadig** [comment]
We might be missing some detail here. I cannot repro this with the latest python extension(`2019.3.6139`) and ptvsd(`4.2.5`):
![image](https://user-images.githubusercontent.com/3840081/55180043-fb2af300-5145-11e9-9ad6-11a05cdc111f.png)

**Comment by int19h** [comment]
I can repro it on Windows. It's not a crash tho, it's Django raising `SystemExit(3)` in the child process to exit it when it needs to reload. Since it's not zero, and it doesn't have a handler, we report it if "Uncaught Exceptions" is checked, per this table:

https://github.com/Microsoft/ptvsd/issues/841#issuecomment-436059489

If debugging is continued with F5, the process exits and reload proceeds normally, with debugger attaching to the new process. 

So basically this just needs some special handling of `SystemExit`. I think our existing table is fine in general, but Django is a special case. There might be others, though, so rather than hardcoding it for Django, maybe we should just provide a new debug option to specify a list of exit codes that aren't errors? And then default it to `[0]` if `BreakOnSystemExitZero` is set, to `[0,3]` if `Django` is set etc.

**Comment by DonJayamanne** [comment]
> . There might be others, though, so rather than hardcoding it for Django, maybe we should just provide a new debug option to specify a list of exit codes that aren't errors?

And possibly ensure the defaults do the job. I.e. if users want to change the default behavior they can provide the necessary settings.

What do you think?

This way we have a consistent behavior with the clients of ptvsd (vs & VSC)

**Comment by DonJayamanne** [comment]
And less work at the other end....😉

**Comment by int19h** [comment]
Turns out that https://github.com/Microsoft/ptvsd/pull/1280 kinda fixed this issue in a different way - since this particular exit is triggered by Django autoreloader, and it's in stdlib, with that fix, we no longer report it as uncaught with `debugStdLib: false` (which is the default).

It is unclear to me whether this is the right behavior, though. The exception originates in library code, but it still propagates through user code (in manage.py in this case) on the way out, and I would expect it to be reported as uncaught when it escapes those, and for the non-zero exit code policies applied at that point.

**Comment by fabioz** [comment]
@int19h the current heuristic to determine if the exception is from the user or from the standard library will only check if the unhandled exception originates from the user or standard library, it doesn't check all the frames it passed through (so, it should be reported as caught once in manage.py but should not be reported as uncaught unless it was actually originated from user code).

Note that it should be straightforward to check the whole stack if that's what we want (related code: `pydevd.PyDB.exclude_exception_by_filter`) -- both reasonings make sense to me, but I'm a bit more inclined to only check the originating location as it is now.

**Comment by int19h** [comment]
The problem with that behavior is that it essentially means that all uncaught exceptions raised from any part of stdlib that's implemented in Python are not considered uncaught - the process just terminates instead of breaking on them. And on the other hand, if you are invoking a builtin method and it raises, then its originating frame is deemed to be user code (since there's no Python frame for native methods), and it is reported as uncaught. So e.g. if you're indexing a list, and the index is invalid, that's reported. But if you're indexing a stdlib collection that's wrapping a list, and it is implemented in Python, then the exception is filtered out, because the originating frame is in stdlib. It seems inconsistent. 

Fundamentally the problem is that the exception can originate from an stdlib frame, but still be triggered explicitly by the user using some stdlib API that is explicitly documented as raising an exception, so they would expect it to be reported.

**Comment by fabioz** [comment]
@int19h I'm ok with that. Can you create a bug report for me to take a look at it?

**Comment by mschoettle** [comment]
I have exactly this problem, without `--noreload` it does not work, breakpoints are just ignored.

```
Version: 1.46.1
Commit: cd9ea6488829f560dc949a8b2fb789f3cdc05f5d
Date: 2020-06-17T21:17:14.222Z
Electron: 7.3.1
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Darwin x64 18.7.0
```

**Comment by int19h** [comment]
Please file a new issue in https://github.com/microsoft/debugpy, and include more detailed repro steps - Python version and distro, terminal log, launch.json etc.
